x/shield/keeper/proposal.go

```diff
@@ -201,38 +196,51 @@ func (k Keeper) RedirectUnbondingEntryToShieldModule(ctx sdk.Context, ubd stakin
 }
 
 // ClaimUnlock unlocks locked collaterals.
-func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID uint64, poolID uint64, loss sdk.Coins) error {
+func (k Keeper) ClaimUnlock(ctx sdk.Context, proposalID uint64, poolID uint64) error {
 	pool, found := k.GetPool(ctx, poolID)
 	if !found {
 		return types.ErrNoPoolFound
 	}
-	lossAmt := loss.AmountOf(k.sk.BondDenom(ctx))
-	pool.TotalCollateral = pool.TotalCollateral.Add(lossAmt)
-	pool.TotalLocked = pool.TotalLocked.Sub(lossAmt)
-	k.SetPool(ctx, pool)
 
-	// update collaterals and providers
+	// Update pool, collaterals and providers.
+	// Take overdraft into considerations.
+	/*
+	 * Example:
+	 * A1 started withdrawing 150: pool.TotalCollateral 400; collateral.Amount 200; provider.Collateral 300.
+	 * Claim proposal 1 locked 200: pool.TotalCollateral 400 --> 200; collateral.Amount 200 --> 100; provider.Collateral 300 --> 200.
+	 * A1 finished withdrawing 150: pool.TotalCollateral 200 --> 100; collateral.Amount 100 --> 0; collateral.Overdraft 0 --> 50; provider.Collateral 200 --> 100.
+	 * A1 deposited 50: pool.TotalCollateral 100 --> 150; collateral.Amount 0 --> 50; collateral.Overdraft 50; provider.Collateral 100 --> 150.
+	 * Claim proposal 1 unlock 200: pool.TotalCollateral 150 --> 300; collateral.Amount 50 --> 100; collateral.Overdraft 50 --> 0; provider.Collateral 150 --> 200.
```

Comment:
So does `Overdraft` become collateralized again if the proposal fails? Shouldn’t it be withdrawn?

Reply:
If 150 is locked while overdraft is 0, then 150 will become collateral again when the proposal fails; if 150 is locked and overdraft is 50, then 100 will become collateral when the proposal fails.

Reply:
Withdraw actually removes collateral. It does not send tokens to users. We have done withdrawal earlier, overdraft only decides how much should be returned when doing unlock.

Reply:
Oh right. The comments make sense. But my confusion also comes from that we only process withdrawal completion based on `validWithdrawAmount` that does not include `collateral.Overdraft`. But then in claim unlock, we're not doing anything with `collateral.Overdraft` other than updating the variable.

Reply:
Oh, we update collaterals accordingly. I guess we have to update `provider.Available` though?

Reply:
We actually update all the collaterals based on `collateral.Overdraft` in claim unlock.

Reply:
Right `provider.Available` should be updated.

Reply:
so many variables to keep track of :D

Reply:
thanks. I think you're right this is the best solution.

Reply:
Thank you!